File: frida_webview_debug.js
/**
 * Frida Hook：强制开启大麦 APP WebView 调试模式
 *
 * 原理：
 *   Android WebView 默认不暴露 DevTools 端口，
 *   需要应用代码调用 WebView.setWebContentsDebuggingEnabled(true)。
 *   大麦 APP 未开启此开关，导致 Chrome DevTools 无法连接（_wd is None）。
 *
 *   本脚本通过 Frida 在运行时强制调用该方法，
 *   使 WebView 在 /proc/net/unix 中暴露 webview_devtools_remote_<pid> 套接字，
 *   从而允许 adb forward localabstract: → 本地 TCP 端口 → CDP 连接。
 *
 * 使用：
 *   方式 1（推荐，spawn 模式，APP 启动前注入）：
 *     frida -U -f cn.damai -l frida_webview_debug.js --no-pause
 *
 *   方式 2（attach 模式，APP 已运行后注入）：
 *     frida -U cn.damai -l frida_webview_debug.js
 *
 *   方式 3（Python 集成，在自动化脚本中调用）：
 *     见下方 frida_webview_debug.py
 *
 * 依赖：
 *   pip install frida frida-tools
 *   手机需运行 frida-server（版本与 frida 一致）
 */

"use strict";

// ─── 颜色输出 ──────────────────────────────────────────────────────────────
var COLOR_RESET = "\x1b[0m";
var COLOR_GREEN = "\x1b[32m";
var COLOR_YELLOW = "\x1b[33m";
var COLOR_CYAN = "\x1b[36m";
var COLOR_RED = "\x1b[31m";

function logInfo(msg) {
    console.log(COLOR_CYAN + "[*] " + msg + COLOR_RESET);
}

function logSuccess(msg) {
    console.log(COLOR_GREEN + "[+] " + msg + COLOR_RESET);
}

function logWarn(msg) {
    console.log(COLOR_YELLOW + "[!] " + msg + COLOR_RESET);
}

function logError(msg) {
    console.log(COLOR_RED + "[-] " + msg + COLOR_RESET);
}

// ─── 核心：强制开启 WebView 调试 ────────────────────────────────────────────

function enableWebViewDebugging() {
    var enabled = false;

    // ── 1. android.webkit.WebView（标准 WebView） ──
    try {
        var WebView = Java.use("android.webkit.WebView");
        WebView.setWebContentsDebuggingEnabled(true);
        logSuccess("android.webkit.WebView.setWebContentsDebuggingEnabled(true) 已调用");
        enabled = true;
    } catch (e) {
        logWarn("android.webkit.WebView hook 失败：" + e);
    }

    // ── 2. com.tencent.smtt.sdk.WebView（腾讯 X5 WebView） ──
    try {
        var X5WebView = Java.use("com.tencent.smtt.sdk.WebView");
        X5WebView.setWebContentsDebuggingEnabled(true);
        logSuccess("com.tencent.smtt.sdk.WebView (X5) 调试已开启");
        enabled = true;
    } catch (e) {
        // 大麦可能不用 X5，忽略
    }

    // ── 3. com.uc.webview.SdkWebView（UC WebView） ──
    try {
        var UCWebView = Java.use("com.uc.webview.SdkWebView");
        UCWebView.setWebContentsDebuggingEnabled(true);
        logSuccess("com.uc.webview.SdkWebView (UC) 调试已开启");
        enabled = true;
    } catch (e) {
        // 忽略
    }

    // ── 4. chromium.webview.ChromeWebView（Chrome Custom Tab） ──
    try {
        var ChromeWebView = Java.use("org.chromium.webview.ChromeWebView");
        ChromeWebView.setWebContentsDebuggingEnabled(true);
        logSuccess("ChromeWebView 调试已开启");
        enabled = true;
    } catch (e) {
        // 忽略
    }

    return enabled;
}

// ─── Hook WebView 构造函数，确保每个实例都开启调试 ──────────────────────────

function hookWebViewConstructors() {
    var hookCount = 0;

    // 标准 WebView 构造函数
    try {
        var WebView = Java.use("android.webkit.WebView");

        // Hook 所有构造函数重载
        var overloads = WebView.$init.overloads;
        for (var i = 0; i < overloads.length; i++) {
            overloads[i].implementation = function () {
                // 先确保全局调试已开启
                Java.use("android.webkit.WebView").setWebContentsDebuggingEnabled(true);
                // 调用原构造函数
                return this.$init.apply(this, arguments);
            };
            hookCount++;
        }
        logSuccess("已 hook android.webkit.WebView 构造函数（" + hookCount + " 个重载）");
    } catch (e) {
        logWarn("WebView 构造函数 hook 失败：" + e);
    }

    // X5 WebView 构造函数
    try {
        var X5WebView = Java.use("com.tencent.smtt.sdk.WebView");
        var x5Overloads = X5WebView.$init.overloads;
        var x5Count = 0;
        for (var j = 0; j < x5Overloads.length; j++) {
            x5Overloads[j].implementation = function () {
                try {
                    Java.use("com.tencent.smtt.sdk.WebView").setWebContentsDebuggingEnabled(true);
                } catch (e) { }
                return this.$init.apply(this, arguments);
            };
            x5Count++;
        }
        if (x5Count > 0) {
            logSuccess("已 hook X5 WebView 构造函数（" + x5Count + " 个重载）");
        }
    } catch (e) {
        // 忽略
    }
}

// ─── Hook setWebContentsDebuggingEnabled 本身，拦截并强制为 true ──────────

function hookSetDebuggingEnabled() {
    try {
        var WebView = Java.use("android.webkit.WebView");

        WebView.setWebContentsDebuggingEnabled.implementation = function (enabled) {
            // 无论传入什么参数，都强制为 true
            logInfo("拦截 setWebContentsDebuggingEnabled(" + enabled + ") → 强制设为 true");
            this.setWebContentsDebuggingEnabled(true);
        };
        logSuccess("已 hook setWebContentsDebuggingEnabled，强制返回 true");
    } catch (e) {
        logWarn("hook setWebContentsDebuggingEnabled 失败：" + e);
    }
}

// ─── 监控 WebView 创建，输出调试信息 ────────────────────────────────────────

function hookWebViewForDiagnostics() {
    try {
        var WebView = Java.use("android.webkit.WebView");

        // Hook loadUrl，记录 WebView 加载的 URL
        WebView.loadUrl.overloads.forEach(function (overload) {
            overload.implementation = function () {
                var url = arguments[0] || "(unknown)";
                logInfo("WebView.loadUrl: " + url);
                return this.loadUrl.apply(this, arguments);
            };
        });
        logSuccess("已 hook WebView.loadUrl 用于诊断");
    } catch (e) {
        logWarn("WebView.loadUrl hook 失败：" + e);
    }
}

// ─── 检查当前已有的 WebView 实例 ────────────────────────────────────────────

function scanExistingWebViews() {
    logInfo("扫描已有的 WebView 实例…");

    try {
        Java.choose("android.webkit.WebView", {
            onMatch: function (instance) {
                var url = "(unknown)";
                try {
                    url = instance.getUrl();
                } catch (e) {
                    try {
                        url = instance.getOriginalUrl();
                    } catch (e2) { }
                }
                logSuccess("发现 WebView 实例：url=" + url);
            },
            onComplete: function () {
                logInfo("WebView 实例扫描完成");
            }
        });
    } catch (e) {
        logWarn("扫描 WebView 实例失败：" + e);
    }
}

// ─── Native 层：Hook DevToolsAgent 启动 ─────────────────────────────────────

function hookNativeDevTools() {
    // 某些 ROM 的 WebView 实现可能在 native 层控制调试开关
    // 尝试 hook chromium 的 DevTools 相关函数

    try {
        // 查找 libwebviewchromium.so 中的调试相关符号
        var modules = Process.enumerateModules();
        var webviewModules = modules.filter(function (m) {
            return m.name.toLowerCase().indexOf("webview") >= 0 ||
                   m.name.toLowerCase().indexOf("chromium") >= 0;
        });

        if (webviewModules.length > 0) {
            logInfo("发现 WebView 相关 native 模块：");
            webviewModules.forEach(function (m) {
                logInfo("  " + m.name + " @ " + m.base + " (size: " + m.size + ")");
            });
        } else {
            logInfo("未发现 WebView native 模块（可能尚未加载）");
        }
    } catch (e) {
        logWarn("Native 模块枚举失败：" + e);
    }
}

// ─── 主入口 ────────────────────────────────────────────────────────────────

logInfo("══════════════════════════════════════════════════");
logInfo("  Frida WebView Debug Hook for cn.damai");
logInfo("  强制开启 WebView 调试模式");
logInfo("══════════════════════════════════════════════════");

Java.perform(function () {
    logInfo("Java VM 已就绪，开始注入…");

    // Step 1: 立即开启全局调试开关
    var ok = enableWebViewDebugging();
    if (ok) {
        logSuccess("WebView 调试模式已全局开启！");
    } else {
        logError("所有 WebView 类均 hook 失败，请确认 APP 已加载 WebView");
    }

    // Step 2: Hook setWebContentsDebuggingEnabled，防止 APP 关闭调试
    hookSetDebuggingEnabled();

    // Step 3: Hook WebView 构造函数，确保新创建的 WebView 也开启调试
    hookWebViewConstructors();

    // Step 4: 诊断 hook（可选，记录 WebView 加载的 URL）
    hookWebViewForDiagnostics();

    // Step 5: 扫描已有的 WebView 实例
    scanExistingWebViews();

    // Step 6: Native 层诊断
    hookNativeDevTools();

    logSuccess("所有 hook 注入完成！WebView DevTools 端口已暴露。");
    logInfo("现在可以在另一个终端运行：");
    logInfo("  adb shell cat /proc/net/unix | grep devtools");
    logInfo("  adb forward tcp:9222 localabstract:webview_devtools_remote_<pid>");
    logInfo("或直接运行 damai_reserve_u2.py 脚本");
});

File: frida_webview_debug.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Frida WebView 调试注入器（Python 集成版）

在自动化脚本中调用，自动注入 Frida hook 开启大麦 APP 的 WebView 调试模式，
无需手动在终端运行 frida 命令。

使用方式 1 — 独立运行（先注入，再运行自动化脚本）：
  python frida_webview_debug.py

使用方式 2 — 在自动化脚本中集成：
  from frida_webview_debug import FridaWebViewDebugHook
  hook = FridaWebViewDebugHook()
  hook.attach("cn.damai")   # attach 到已运行的 APP
  # 或
  hook.spawn("cn.damai")    # spawn 模式（APP 启动前注入）

使用方式 3 — 作为上下文管理器：
  with FridaWebViewDebugHook.spawn("cn.damai") as hook:
      # 在此期间 WebView 调试已开启
      run_automation()
  # 退出时自动 detach

依赖：
  pip install frida frida-tools
  手机需运行 frida-server（版本与本地 frida 一致）
"""

from __future__ import annotations

import os
import sys
import time
import threading
from pathlib import Path
from typing import Optional

try:
    import frida
except ImportError:
    sys.stderr.write(
        "缺少依赖 frida，请先执行：pip install frida frida-tools\n"
    )
    sys.exit(1)


# ─── Frida JS 脚本路径 ──────────────────────────────────────────────────────

SCRIPT_DIR = Path(__file__).parent
FRIDA_JS_PATH = SCRIPT_DIR / "frida_webview_debug.js"


# ─── 核心 Hook 类 ────────────────────────────────────────────────────────────

class FridaWebViewDebugHook:
    """Frida WebView 调试注入器。"""

    def __init__(self, verbose: bool = False):
        """
        Args:
            verbose: 是否输出详细日志
        """
        self.verbose = verbose
        self._device: Optional[frida.core.Device] = None
        self._session: Optional[frida.core.Session] = None
        self._script: Optional[frida.core.Script] = None
        self._pid: Optional[int] = None
        self._messages: list = []
        self._attached = False

    def _log(self, msg: str) -> None:
        if self.verbose:
            print(f"  [frida] {msg}")

    @staticmethod
    def _find_device() -> frida.core.Device:
        """查找 USB 连接的 Android 设备。"""
        try:
            device = frida.get_usb_device(timeout=5)
            return device
        except Exception:
            pass

        # 备用：尝试远程设备
        try:
            device = frida.get_remote_device()
            return device
        except Exception:
            pass

        # 列出所有设备
        devices = frida.enumerate_devices()
        usb_devices = [d for d in devices if d.type == "usb"]
        if usb_devices:
            return usb_devices[0]

        raise RuntimeError(
            "未找到 USB 设备。请确认：\n"
            "  1. 手机已通过 USB 连接电脑\n"
            "  2. 手机已开启 USB 调试\n"
            "  3. 手机上已运行 frida-server\n"
            "  4. frida-server 版本与本地 frida 一致"
        )

    def _load_js_script(self) -> str:
        """加载 Frida JS 脚本。"""
        if FRIDA_JS_PATH.is_file():
            return FRIDA_JS_PATH.read_text(encoding="utf-8")
        else:
            # 内嵌精简版脚本
            return self._get_inline_script()

    @staticmethod
    def _get_inline_script() -> str:
        """内嵌的精简版 Frida JS 脚本（当外部 JS 文件不存在时使用）。"""
        return r"""
"use strict";

function enableWebViewDebugging() {
    var enabled = false;
    var classes = [
        "android.webkit.WebView",
        "com.tencent.smtt.sdk.WebView",
        "com.uc.webview.SdkWebView",
    ];
    for (var i = 0; i < classes.length; i++) {
        try {
            var Cls = Java.use(classes[i]);
            Cls.setWebContentsDebuggingEnabled(true);
            console.log("[+] " + classes[i] + " 调试已开启");
            enabled = true;
        } catch (e) {}
    }
    return enabled;
}

Java.perform(function () {
    // 立即开启
    enableWebViewDebugging();

    // Hook setWebContentsDebuggingEnabled 防止关闭
    try {
        var WebView = Java.use("android.webkit.WebView");
        WebView.setWebContentsDebuggingEnabled.implementation = function (v) {
            this.setWebContentsDebuggingEnabled(true);
        };
    } catch (e) {}

    // Hook 构造函数
    try {
        var WebView2 = Java.use("android.webkit.WebView");
        WebView2.$init.overloads.forEach(function (overload) {
            overload.implementation = function () {
                Java.use("android.webkit.WebView").setWebContentsDebuggingEnabled(true);
                return this.$init.apply(this, arguments);
            };
        });
    } catch (e) {}

    console.log("[+] Frida WebView Debug Hook 注入完成");
});
"""

    def _on_message(self, message: dict, data: Optional[bytes]) -> None:
        """处理 Frida 脚本消息。"""
        self._messages.append(message)

        if message["type"] == "send":
            payload = message.get("payload", "")
            print(f"  [frida→py] {payload}")
        elif message["type"] == "error":
            stack = message.get("stack", "")
            description = message.get("description", "")
            print(f"  [frida error] {description}\n{stack}", file=sys.stderr)
        else:
            self._log(f"消息：{message}")

    def attach(self, package_name: str = "cn.damai") -> bool:
        """Attach 到已运行的 APP 进程。

        Args:
            package_name: APP 包名

        Returns:
            是否成功注入
        """
        print(f"🔌 正在 attach 到 {package_name}…")

        try:
            self._device = self._find_device()
            print(f"✅ 已连接设备：{self._device.name} (id={self._device.id})")
        except RuntimeError as e:
            print(f"❌ {e}")
            return False

        # 查找目标进程
        try:
            process = self._device.get_process(package_name)
            self._pid = process.pid
            print(f"✅ 找到进程：{process.name} (pid={process.pid})")
        except frida.ProcessNotFoundError:
            print(f"❌ 未找到进程 {package_name}，请先启动 APP")
            return False

        # Attach
        try:
            self._session = self._device.attach(self._pid)
            self._log(f"已 attach 到 pid={self._pid}")
        except Exception as e:
            print(f"❌ attach 失败：{e}")
            return False

        # 注入脚本
        return self._inject_script()

    def spawn(self, package_name: str = "cn.damai") -> bool:
        """Spawn 模式：在 APP 启动前注入（更可靠）。

        Args:
            package_name: APP 包名

        Returns:
            是否成功注入
        """
        print(f"🚀 正在 spawn {package_name}…")

        try:
            self._device = self._find_device()
            print(f"✅ 已连接设备：{self._device.name} (id={self._device.id})")
        except RuntimeError as e:
            print(f"❌ {e}")
            return False

        # Spawn
        try:
            self._pid = self._device.spawn([package_name])
            print(f"✅ 已 spawn 进程 (pid={self._pid})")
        except Exception as e:
            print(f"❌ spawn 失败：{e}")
            print("请确认：")
            print("  1. frida-server 版本与本地 frida 一致")
            print("  2. 手机上有足够权限运行 frida-server（需 root）")
            return False

        # Attach
        try:
            self._session = self._device.attach(self._pid)
            self._log(f"已 attach 到 pid={self._pid}")
        except Exception as e:
            print(f"❌ attach 失败：{e}")
            return False

        # 注入脚本
        if not self._inject_script():
            return False

        # Resume APP
        try:
            self._device.resume(self._pid)
            print(f"✅ APP 已 resume，WebView 调试已开启")
        except Exception as e:
            self._log(f"resume 失败（非致命）：{e}")

        return True

    def _inject_script(self) -> bool:
        """注入 Frida JS 脚本。"""
        js_code = self._load_js_script()
        self._log(f"JS 脚本长度：{len(js_code)} 字符")

        try:
            self._script = self._session.create_script(js_code)
            self._script.on("message", self._on_message)
            self._script.load()
            self._attached = True
            print("✅ Frida hook 脚本已注入")
            return True
        except Exception as e:
            print(f"❌ 脚本注入失败：{e}")
            return False

    def detach(self) -> None:
        """断开 Frida 连接。"""
        if self._script:
            try:
                self._script.unload()
            except Exception:
                pass
            self._script = None

        if self._session:
            try:
                self._session.detach()
            except Exception:
                pass
            self._session = None

        self._attached = False
        self._log("已 detach")

    def is_attached(self) -> bool:
        """是否已注入。"""
        return self._attached

    def wait_until_detached(self) -> None:
        """阻塞等待，直到 Frida 连接断开（APP 退出或手动 detach）。"""
        if not self._session:
            return

        print("⏳ Frida hook 已激活，按 Ctrl+C 退出…")
        try:
            # 保持 session 活跃
            while self._attached:
                time.sleep(1)
        except KeyboardInterrupt:
            print("\n正在断开…")
            self.detach()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.detach()
        return False


# ─── 辅助函数 ────────────────────────────────────────────────────────────────

def check_frida_server() -> bool:
    """检查设备上 frida-server 是否运行。"""
    try:
        device = frida.get_usb_device(timeout=3)
        # 尝试枚举进程，如果 frida-server 没运行会超时
        processes = device.enumerate_processes()
        frida_procs = [p for p in processes if "frida" in p.name.lower()]
        if frida_procs:
            print(f"✅ frida-server 已运行：{frida_procs}")
            return True
        else:
            print("⚠️ 未检测到 frida-server 进程")
            return False
    except Exception as e:
        print(f"❌ 无法连接设备：{e}")
        return False


def inject_and_wait(package_name: str = "cn.damai", spawn_mode: bool = True) -> None:
    """注入 Frida hook 并保持运行。

    Args:
        package_name: APP 包名
        spawn_mode: True=spawn 模式（推荐），False=attach 模式
    """
    hook = FridaWebViewDebugHook(verbose=True)

    if spawn_mode:
        ok = hook.spawn(package_name)
    else:
        ok = hook.attach(package_name)

    if not ok:
        print("\n❌ 注入失败。")
        print("请确认：")
        print("  1. 手机已 root 且已运行 frida-server")
        print("  2. frida-server 版本与本地 frida 一致")
        print("     检查方法：")
        print("       pip show frida          # 查看本地版本")
        print("       adb shell /data/local/tmp/frida-server --version  # 查看设备版本")
        print("  3. 如使用 Magisk，可安装 MagiskFrida 模块自动启动 frida-server")
        sys.exit(1)

    print()
    print("═" * 50)
    print("  WebView 调试已开启！")
    print("  现在可以在另一个终端运行自动化脚本：")
    print(f"    python damai_reserve_u2.py --verbose")
    print("═" * 50)
    print()

    # 保持 hook 活跃
    hook.wait_until_detached()


# ─── 主入口 ──────────────────────────────────────────────────────────────────

def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(
        description="Frida WebView 调试注入器 — 强制开启大麦 APP WebView DevTools",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=(
            "示例:\n"
            "  python frida_webview_debug.py                           # spawn 模式（推荐）\n"
            "  python frida_webview_debug.py --attach                  # attach 模式\n"
            "  python frida_webview_debug.py --package cn.damai        # 指定包名\n"
            "  python frida_webview_debug.py --check                   # 仅检查 frida-server\n"
            "\n"
            "前置条件:\n"
            "  1. 手机已 root\n"
            "  2. 已安装 frida-server 且版本与本地 frida 一致\n"
            "     pip install frida frida-tools\n"
            "     adb push frida-server /data/local/tmp/\n"
            "     adb shell 'chmod 755 /data/local/tmp/frida-server'\n"
            "     adb shell '/data/local/tmp/frida-server &'\n"
        ),
    )
    parser.add_argument("--package", type=str, default="cn.damai",
                        help="APP 包名（默认 cn.damai）")
    parser.add_argument("--attach", action="store_true",
                        help="使用 attach 模式（默认 spawn 模式）")
    parser.add_argument("--check", action="store_true",
                        help="仅检查 frida-server 是否运行")
    parser.add_argument("-v", "--verbose", action="store_true",
                        help="详细日志")
    args = parser.parse_args()

    # Windows 控制台 UTF-8
    try:
        sys.stdout.reconfigure(encoding="utf-8")
        sys.stderr.reconfigure(encoding="utf-8")
    except Exception:
        pass

    print("╔════════════════════════════════════════════════╗")
    print("║  Frida WebView Debug Hook                       ║")
    print("║  强制开启大麦 APP WebView 调试模式              ║")
    print("╚════════════════════════════════════════════════╝")
    print()

    # 检查 frida 版本
    print(f"📦 frida 版本：{frida.__version__}")

    if args.check:
        check_frida_server()
        return

    inject_and_wait(
        package_name=args.package,
        spawn_mode=not args.attach,
    )


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n已退出。")
